### PR TITLE
re3 source lol

### PR DESCRIPTION
**Pull request by anarh1st47**
If you are looking to file or dispute a takedown notice by posting to this repository, please STOP :stop_sign: because we do not accept Pull Requests or other contributions to this repository.

Read on to learn about the available paths forward.

Please note that re-posting the exact same content that was the subject of a takedown notice without following the proper process ([outlined below](#responding-to-a-dmca-notice)) is a violation of GitHub’s [DMCA Policy](https://docs.github.com/en/github/site-policy/dmca-takedown-policy) and [Terms of Service](https://docs.github.com/en/github/site-policy/github-acceptable-use-policies). If you commit or post content to this repository that violates our Terms of Service, we will delete that content and may suspend access to your account as well.

#### Submitting a DMCA Notice

If you are a copyright owner wishing to submit a takedown notice, read our [DMCA Policy](https://docs.github.com/en/free-pro-team@latest/github/site-policy/dmca-takedown-policy) and [Guide to Submitting a DMCA Takedown Notice](https://docs.github.com/en/free-pro-team@latest/github/site-policy/guide-to-submitting-a-dmca-takedown-notice). You can submit the actual notice using our special [Copyright Claims Contact Form](https://github.com/contact/dmca).

#### Responding to a DMCA Notice

If you are the owner of a repository that has been taken down, you have two main options:

  - Do you want to [make changes](https://docs.github.com/en/free-pro-team@latest/github/site-policy/dmca-takedown-policy#c-what-if-i-inadvertently-missed-the-window-to-make-changes) to the repository that would remove the allegedly infringing content? If that is possible in your case, [contact us](https://support.github.com/contact) to let us know that you would like to make the changes.

  - Do you want to formally dispute the action by [submitting a counter notice](https://docs.github.com/en/free-pro-team@latest/github/site-policy/guide-to-submitting-a-dmca-counter-notice)? Maybe the person sending the takedown notice does not hold the copyright or did not realize that you have a license or made some other mistake in their takedown notice. If you believe your content on GitHub was mistakenly disabled by a DMCA takedown request, you have the right to contest the takedown by [submitting a counter notice](https://docs.github.com/en/free-pro-team@latest/github/site-policy/guide-to-submitting-a-dmca-counter-notice). If you do, we will wait 10-14 days and then re-enable your content unless the copyright owner initiates a legal action before then.

If you do not want to make changes or dispute the notice, but still have general concerns about the copyright laws and how they apply in your case, know that GitHub and developers have the opportunity and a voice to advocate for changes in law and public policy to better support software development. We are constantly looking to [advocate for developers](https://github.blog/category/company/policy/), so feel free to [reach out](https://support.github.com/contact) and let us know your concerns. We also encourage you to [learn more about copyright and speak up](https://docs.github.com/en/free-pro-team@latest/github/site-policy/dmca-takedown-policy#learn-more-and-speak-up) by reaching out to the Copyright Office or your local lawmakers to voice your concerns.

### CODE REVIEW

**Comment by ishlun** [comment]
xD

**Comment by anarh1st47** [comment]
reverse engineering is not a crime

**Comment by ishlun** [comment]
completely agree

**Review by StraussT2 — APPROVED** [review]

**Comment by BoredAlchemist** [comment]
fooking micro$oft, always knew github would go to shit once M$ bought it, people should migrate their code to Gitlab

**Comment by Aymen545** [comment]
not cool man not cool

**Comment by username0x0a** [comment]
As an analogy to Linus vs Nvidia – Rockstar, f••k you! What's the point of fighting against an initiative that can only increase your own product purchases? 🤔 I would surely buy your game on Steam again, just so I can run it easily on a Mac, which is impossible at this moment.

Think before doing stupid things!

**Comment by RossComputerGuy** [comment]
So is nouveau source code not allowed on GitHub anymore?

**Comment by alula** [comment]
> fooking micro$oft, always knew github would go to shit once M$ bought it, people should migrate their code to Gitlab

It's not GitHub's fault, every US-based service HAS to comply with laws. Removal of re3 was a legal requirement and any dispute has to be handled later.

**Comment by improvised-explosive-device** [comment]
Best of luck to the re3 team, both 3&VC work pretty flawlessly and I would hate to see them be forced to abandon the project.

> > fooking micro$oft, always knew github would go to shit once M$ bought it, people should migrate their code to Gitlab
> 
> It's not GitHub's fault, every US-based service HAS to comply with laws. Removal of re3 was a legal requirement and any dispute has to be handled later.

@alula No logic allowed here, get out plz.

**Comment by BoredAlchemist** [comment]
> > fooking micro$oft, always knew github would go to shit once M$ bought it, people should migrate their code to Gitlab
> 
> @BoredAlchemist It's not GitHub's fault, every US-based service HAS to comply with laws. Removal of re3 was a legal requirement and any dispute has to be handled later.

they didn't use any rockstar code did they as far as i am aware

**Comment by RossComputerGuy** [comment]
As I'm aware, RE3 is a reverse engineered program. Meaning the code is original to the project itself and it is legal to have it. This is not like stealing a painting and claiming it your own. It's more like making a painting based off of someone else's but in your own style. It's your original work but it is based off something. The same can be applied to videos, music, etc. If RE3 can't exist then a lot of other projects like nouveua, wine, ReactOS, etc.

**Comment by improvised-explosive-device** [comment]
> As I'm aware, RE3 is a reverse engineered program. Meaning the code is original to the project itself and it is legal to have it. This is not like stealing a painting and claiming it your own. It's more like making a painting based off of someone else's but in your own style. It's your original work but it is based off something. The same can be applied to videos, music, etc. If RE3 can't exist then a lot of other projects like nouveua, wine, ReactOS, etc.

One of the devs uploaded a video of themselves reversing a function in IDA. 🤷

**Comment by RusJJ** [comment]
> > @BoredAlchemist: they didn't use any rockstar code did they as far as i am aware
> 
> > @RossComputerGuy: As I'm aware, RE3 is a reverse engineered program. Meaning the code is original to the project itself and it is legal to have it. This is not like stealing a painting and claiming it your own. It's more like making a painting based off of someone else's but in your own style. It's your original work but it is based off something. The same can be applied to videos, music, etc. If RE3 can't exist then a lot of other projects like nouveua, wine, ReactOS, etc.
> 
> One of the devs uploaded a video of themselves reversing a function in IDA. 🤷

So what? Reverse Engineered code -> Not the same as the original one.

**Comment by Ornias1993** [comment]
Any links to non-github copies of the source?

**Comment by AsciiWolf** [comment]
> reverse engineering is not a crime

Yeah, just like with Linux drivers made using reverse engineering, unofficial, open source, game engine recreations **should be considered fair use** if they still require original game data, and sure enough, there is already a lot of [unofficial open source engines](https://en.wikipedia.org/wiki/List_of_game_engine_recreations) and many of them were made using reverse engineering.

**Comment by improvised-explosive-device** [comment]
> So what? Reverse Engineered code -> Not the same as the original one.

https://wiki.winehq.org/Clean_Room_Guidelines
There is a reason why the Wine developers are so strict about this.

**Review by HRudyPlayZ — APPROVED** [review]

**Comment by Kein** [comment]
> reverse engineering is not a crime

it is not for you to decide but for big money and big corporations. You exist only to consume and follow the law they dictate.

**Comment by normanlol** [comment]
> > reverse engineering is not a crime
> it is not for you to decide but for big money and big corporations. You exist only to consume and follow the law they dictate.

it isn't a crime in the slightest. it is legally allowed as long as they don't use code owned by take-two

**Comment by Kein** [comment]
Good luck having enough money to win the case if you decide to defend your "legal right".

**Comment by ZReC** [comment]
If this was a troll, jokes on you, I will be actually relieved

**Comment by lei-hsia** [comment]
I'm the owner of this repo. Technically, it's more like a personal notes than a repo. I'm pleased that many benefited from my notes in the past two years. 

I don't quite understand why my repo was taken down, though, for 3 reasons: 

1. 
I paid for Educative's video courses, made notes and posted on Github. Surely the notes cannot be 100% the same as video courses because anyone would add his/her own thoughts. Also, I've put "Source: Educative course" with a URL in the forefront: I'm making free ads for them. And I'm called a piracy. Seriously, how could it be that some notes are a piracy of video courses?

2. 
Their lawyer claimed "I have reviewed a REPRESENTATIVE sample of the forked repositories and conclude that the following members within this network are allegedly infringing on our content". 

Technically speaking, anything called "representative" are de facto not. Anyone with little probabilities theory knowledge would know this. I don't know how they define representative. 

I would guess they see something similar with a predefined biased mindset, randomly picked one, and claimed a piracy. But all System Design prep materials are similar. As a result, I doubt how they decide a piracy; 25% similar, 30%, or 35%? Beyond me.

3. 
I have many copies of my notes. I can post anywhere. I see no point in taking my notes down here. Maybe I will post it on Facebook, or Google Docs tomorrow.

I think I'll file a counter notice on weekends.

**Comment by jessy-lua** [comment]
github is be dying
taking DCMA from trolls without actually analysing it

**Comment by RossComputerGuy** [comment]
Or the Linux kernel itself because it "has Unix code" /s

File: src/core/PlayerInfo.cpp [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+#ifdef VC_PED_PORTS
+						if (carBelow->GetStatus() != STATUS_WRECKED && carBelow->GetUp().z > 0.3f)
+#else
+						if (carBelow->GetStatus() != STATUS_WRECKED)
+#endif
+							m_pPed->SetSeekBoatPosition(carBelow);
+					}
+				}
+				// Find closest car
+				if (!weAreOnBoat) {
```
Comment (swift502) [inline src/core/PlayerInfo.cpp:192]: I'm thinking if `weAreOnBoat` is true, maybe Take-Two shouldn't frivolously take down community projects which fix and improve their terribly maintained games.

Just a thought.

**Comment by Apollyon094** [comment]
o7

**Comment by NameIsA** [comment]
> Or the Linux kernel itself because it "has Unix code" /s

That's a strawman fallacy.

**Review by BatuAksoy — APPROVED** [review]

**Review by ZReC — APPROVED** [review]

**Comment by Jan200101** [comment]
> One of the devs uploaded a video of themselves reversing a function in IDA. shrug

If this is true then it is entirely legal for Rockstar or TakeTwo to take re3 down.

Essentially what has been done with re3 is that the entire game has been picked apart and put back into its original pieces as has been done countless done before with [sm64](https://github.com/n64decomp/sm64) for example.

Those bits being picked apart? they are still owned by Rockstar and picking them apart is infact NOT fair use.

For re3 to have a legal leg to stand on the developers must have reverse engineered GTA 3 using the [clean room design](https://en.wikipedia.org/wiki/Clean_room_design) or to put into more understandable words:
instead of taking the actual game apart someone looks at how the game functions and documents it
then someone who has never looked at the executable in a disassembler takes that documentation and rebuilds the game
its incredible hard and expensive to do which is why not many do it.

Let me repeat this for the children in the back

THE TAKEDOWN WAS ENTIRELY LEGAL AND THERE IS NOTHING YOU CAN DO

**Comment by swift502** [comment]
> THERE IS NOTHING YOU CAN DO

Yeah no kidding.

**Comment by PizzaehTheHuett** [comment]
> > One of the devs uploaded a video of themselves reversing a function in IDA. shrug
> 
> If this is true then it is entirely legal for Rockstar or TakeTwo to take re3 down.
> 
> Essentially what has been done with re3 is that the entire game has been picked apart and put back into its original pieces as has been done countless done before with [sm64](https://github.com/n64decomp/sm64) for example.
> 
> Those bits being picked apart? they are still owned by Rockstar and picking them apart is infact NOT fair use.
> 
> For re3 to have a legal leg to stand on the developers must have reverse engineered GTA 3 using the [clean room design](https://en.wikipedia.org/wiki/Clean_room_design) or to put into more understandable words:
> instead of taking the actual game apart someone looks at how the game functions and documents it
> then someone who has never looked at the executable in a disassembler takes that documentation and rebuilds the game
> its incredible hard and expensive to do which is why not many do it.
> 
> Let me repeat this for the children in the back
> 
> THE TAKEDOWN WAS ENTIRELY LEGAL AND THERE IS NOTHING YOU CAN DO

Not so fast there. Let me explain for those who don't understand how these things work.

Actually there's alot that can be done and already has. You can send all the DMCA requests and lawyers you want and you even managed to get it removed from GitHub, but the source code is now all over the internet. It is yourselves who are powerless here and you will never get that cat back in the bag.

So it dosen't get reinstated on Github. No big deal. Lol. The internet will happily accept it into permanent custody.

**Comment by Jan200101** [comment]
@PizzaehTheHuett 
It is true that once on the internet it will always be there
But re3 will never be able to continue development on any public git service so the project is effectively halted

**Comment by vitalkanev** [comment]
CC to anyone who works for/at GitHub to close or remove this pull request. Thank you.

**Comment by AsciiWolf** [comment]
@Jan200101 Actually, even if this is true, the development of re3 **can continue** on any git service hosted in Europe, by anyone living in Europe, since, as far as I know, reverse engineering is not considered a copyright violation there.

**Comment by Jan200101** [comment]
> @Jan200101 Actually, even if this is true, the development of re3 **can continue** on any git service hosted in Europe, by anyone living in Europe, since, as far as I know, reverse engineering is not considered a copyright violation there.

You are not understanding copyright law
re3 is illegal in the US and EU

**Comment by pkubaj** [comment]
LGTM, please merge!

**Comment by Jan200101** [comment]
> You are not understanding copyright law
> re3 is illegal in the US and EU

this is getting downvoted and I believe there is a lack of understanding (or people are just mad they can't have their way)
the way reverse engineering was done here is by looking at the .exe
the binary, the executable and seeing what it does exactly
the game was written in C++ and anyone with enough knowledge of the C++ standard and the compiler which was used (info about which can be found in the exe itself since the compiler embeds that) and the raw 0's and 1's are made back into code.

This isn't legal
This would be like listening to a song with headphones, playing that exact song and then recording and distributing it.
Sure its you playing it but the actual thing it was derived from is not yours and just playing it does give you permission to do as you please.

Europe is not different than America about this, people who own something are able to defend their property from theft.

> @Jan200101 Actually, even if this is true, the development of re3 **can continue** on any git service hosted in Europe, by anyone living in Europe, since, as far as I know, reverse engineering is not considered a copyright violation there.

and following up on this again
yes development of re3 can continue by a small collective of people with a private git host
but not in the same way that development was done previously
console ports will especially suffer from this

**Comment by AsciiWolf** [comment]
@Jan200101: Is the re3 code an (almost) exact copy of the original? If so, then I suppose that you are probably right (although reverse engineering is not illegal in EU, so this is still more like a grey area).

**Comment by Jan200101** [comment]
> 
> 
> @Jan200101: Is the re3 code an (almost) exact copy of the original? If so, then I suppose that you are probably right (although reverse engineering is not illegal in EU, so I think that this is still more like a grey area, but I may be wrong).

Reverse Engineering is not Illegal in the US or EU if following the Clean Room design.
There have been many court cases like [Sega v. Accoldate](https://en.wikipedia.org/wiki/Sega_v._Accolade) that can confirm this.

the problem is that the re3 code was gained by disassembling the .exe that Rockstar owns, anything derived from that is essentially Rockstars property

**Comment by BatuAksoy** [comment]
@Jan200101 I wonder how hard is proving it is not clean room. Can we tell it only by looking at the code?

Anyway, there will be a counter notice im sure.

**Comment by Jan200101** [comment]
> @Jan200101 I wonder how hard is proving it is not clean room. Can we tell it only by looking at the code?
> 
> Anyway, there will be a counter notice im sure.

If what was previously stated, that there is footage of a developer opening GTA3 with IDA, is true then that alone is enough prove that it was notr done clean room

**Comment by BatuAksoy** [comment]
Well then they can remove the code which can be proved as "not" clean room and reactivate the repo claiming rest is clean room right?

**Comment by Jan200101** [comment]
> 
> 
> Well then they can remove the code which can be proved as "not" clean room and reactivate the repo claiming rest is clean room right?

well yes
but in github's case it it takes the explosive reaction of disabling an entire repository from public viewing
this isn't because github is ignorant but because Github is a business that needs to protect itself, acting any differently can easily end with companies going after github.

If we go with the narative that re3 was done entirely clean room (I am not saying its not, I am not invested enough in this situation to know) then yes the developers have a legal leg to stand on and can appeal the DMCA request
from there on the company taking action will probably sue against the developers of re3 which takes a lot of money and effort and will inevitable set a new precedent for people doing the same

**Comment by BatuAksoy** [comment]
I understand the case ty for the info. From a different point of view, imo nobody can stop open source projects, which may sound illegal. Even here, in github there are other mirror repositories of re3, which will require extra effort to keep them away. So if people want to keep re3 in github they will.

**Comment by sleepytsu** [comment]
h

**Comment by HRudyPlayZ** [comment]
Keep in mind that decompiling an app isn't technically illegal. What is though, is reusing/reuploading/ whatever the copyrighted code, unless proven otherwise, re3 wasn't reusing the copyrighted code. As for copyright laws, DMCA doesn't apply in Europe indeed, instead there's technically the EUCD which basically is never applied but still exists. Keep in mind that in a case like this, it doesn't matter which country the repo author is from, unless they directly go to court, it's the country where Github resides
that's relevant. Github is in the US and as such is affected by DMCA, so either if you're in another country, as long as you post on Github, you can get DMCAed.
I don't think this was a legal DMCA as there's no copyrighted assets (this includes code) bundled in. Though in this case, they probably don't care about legality as they have the money to enforce anything. I'm not a lawyer so don't take any legal advice from me though.

**Comment by Jan200101** [comment]
> 
> 
> Keep in mind that decompiling an app isn't technically illegal. What is though, is reusing/reuploading/ whatever the copyrighted code, unless proven otherwise, re3 wasn't reusing the copyrighted code. As for copyright laws, DMCA doesn't apply in Europe indeed, instead there's technically the EUCD which basically is never applied but still exists. Keep in mind that in a case like this, it doesn't matter which country the repo author is from, unless they directly go to court, it's the country where Github resides
> that's relevant. Github is in the US and as such is affected by DMCA, so either if you're in another country, as long as you post on Github, you can get DMCAed.

If a previous statement in this threat is true and it can be proven that a developer for re3 used IDA to disassemble the original binary then that is enough to get any commit from that developer shut down on the grounds that the developer is either reusing the disassembled output or something derived from it.

Yes DMCA does not apply in Europe and an equivelent would be the EUCD, as you mentioned.
The reason you see less EUCD claims than DMCA is because most companies or groups to business in the US, like Github as you mentioned, making it pointless to file a EUCD if a DMCA will do the same job and essentially cut off most people from whatever was hit.

> I don't think this was a legal DMCA as there's no copyrighted assets (this includes code) bundled in. Though in this case, they probably don't care about legality as they have the money to enforce anything. I'm not a lawyer so don't take any legal advice from me though.

Let me approach this in a different way
The [Super Mario 64](https://github.com/n64decomp/sm64) decompile does not contain any textures for the game.
So its legal right? are the models and geometry nintendos property? What about all the ingame text that hard to be written and translated?

While re3 does not contain models and so on because the game simply uses a reusable format that can be loaded in, it still contains the raw and processed translation files.
Is that not Rockstars property? is the hard work and money that was put into this not worth their ownership?

**Comment by ZReC** [comment]
Don't worry it may be solved, there's still hopes. No C&D was sent so far, it's just this menacing dmca

**Comment by ZReC** [comment]
Don't worry it may be solved, there's still hope. No C&D was sent so far, it's just this menacing dmca

**Comment by ItzSwirlz** [comment]
not cool man

**Comment by monyarm** [comment]
> > One of the devs uploaded a video of themselves reversing a function in IDA. shrug
> 
> If this is true then it is entirely legal for Rockstar or TakeTwo to take re3 down.
> 
> Essentially what has been done with re3 is that the entire game has been picked apart and put back into its original pieces as has been done countless done before with [sm64](https://github.com/n64decomp/sm64) for example.
> 
> Those bits being picked apart? they are still owned by Rockstar and picking them apart is infact NOT fair use.
> 
> For re3 to have a legal leg to stand on the developers must have reverse engineered GTA 3 using the [clean room design](https://en.wikipedia.org/wiki/Clean_room_design) or to put into more understandable words:
> instead of taking the actual game apart someone looks at how the game functions and documents it
> then someone who has never looked at the executable in a disassembler takes that documentation and rebuilds the game
> its incredible hard and expensive to do which is why not many do it.
> 
> Let me repeat this for the children in the back
> 
> THE TAKEDOWN WAS ENTIRELY LEGAL AND THERE IS NOTHING YOU CAN DO

This couldn't be farther from the truth.
If this was true, and using IDA/Ghidra + a debugger to reverse engineer an executable wasn't allowed, then every other reverse engineered game engine like OpenMW or ScummVM would be illegal.

**Comment by Jan200101** [comment]
> This couldn't be farther from the truth.
> If this was true, and using IDA/Ghidra + a debugger to reverse engineer an executable wasn't allowed, then every other reverse engineered game engine like OpenMW or ScummVM would be illegal.

as it stands OpenMW is a recreation that aims to be compatible with Gamebryo.
ScummVM is a recreation of game engines that used the SCUMM system and so can interpret the scripting language
neither of these two projects aim to be decompilations or similar and simply interpret the game data which is people are free to do.
[devilution](https://github.com/diasurgical/devilution) would be an example for something that violates copyright law

**Comment by monyarm** [comment]
ScummVM and a number of its engines were literally reverse engineered
through IDA. They even have an entire page on their wiki explaining how to
use IDA and the Dosbox Debugger in order to reverse engineer a game.

**Comment by hackermondev** [comment]
does sending a message here alert everyone

**Comment by RusJJ** [comment]
> does sending a message here alert everyone

Yes.

**Comment by 0x416c69** [comment]
Well I don't know about the laws in US and EU but, if you guys are looking for a public git hosting without the problems of DMCA, EUCD, GFYS and STBA, I can host this on a proud Iranian server, have fun sending takedown notices here t2 🤣

**Comment by jessy-lua** [comment]
https://torrentfreak.com/github-restores-reverse-engineered-gta-code-following-dmca-counter-notice-210507/

So when DMCA will be disabled for officially repository.??

**Comment by Jan200101** [comment]
> 
> 
> https://torrentfreak.com/github-restores-reverse-engineered-gta-code-following-dmca-counter-notice-210507/
> 
> So when DMCA will be disabled for officially repository.??

once the official repository sends a counter-claim

**Comment by paulsapps** [comment]
@Jan200101 decompiling C++ is nothing like C# or other "managed" languages. To try to say that using IDA to "decompile" C++ is illegal is pure nonsense. They haven't seen any of the original code therefore it is not possible that it is copyright infringement of the original code. Otherwise any rewrite that happens to produce a clone engine using the original files could easily be deemed illegal for the same reasons. Also RE3 isn't being sold for profit either.

**Comment by CyberFlameGO** [comment]
Did this dude just PR the dmca repo with the code that was taken down?

**Pull request closed without merging**